homeassistant/components/sonos/media_browser.py

```diff
@@ -490,15 +499,45 @@ def get_content_id(item: DidlObject) -> str:
 def get_media(
     media_library: MusicLibrary, item_id: str, search_type: str
 ) -> MusicServiceItem:
-    """Fetch media/album."""
+    """Fetch a single media/album."""
+    _LOGGER.debug("get_media item_id [%s], search_type [%s]", item_id, search_type)
     search_type = MEDIA_TYPES_TO_SONOS.get(search_type, search_type)
 
     if not item_id.startswith("A:ALBUM") and search_type == SONOS_ALBUM:
         item_id = "A:ALBUMARTIST/" + "/".join(item_id.split("/")[2:])
 
-    search_term = item_id.split("/")[-1]
-    matches = media_library.get_music_library_information(
-        search_type, search_term=search_term, full_album_art_uri=True
+    if item_id.startswith("A:ALBUM/") or search_type == "tracks":
+        search_term = urllib.parse.unquote(item_id.split("/")[-1])
+        matches = media_library.get_music_library_information(
+            search_type, search_term=search_term, full_album_art_uri=True
+        )
+    else:
+        # When requesting media by album_artist, composer, genre use the browse interface
+        # to navigate the hierarchy. This occurs when invoked from media browser or service
+        # calls
+        # Example: A:ALBUMARTIST/Neil Young/Greatest Hits - get specific album
+        # Example: A:ALBUMARTIST/Neil Young - get all albums
+        # Others: composer, genre
+        # A:<topic>/<name>/<optional title>
+        splits = item_id.split("/")
+        title = urllib.parse.unquote(splits[2]) if len(splits) > 2 else None
+        browse_id_string = splits[0] + "/" + splits[1]
+        matches = media_library.browse_by_idstring(
+            search_type, browse_id_string, full_album_art_uri=True
+        )
+        if title:
+            result = next(
+                (item for item in matches if (title == item.title)),
+                None,
+            )
+            matches = [result]
```

Comment:
The construct ends up working out, since matches =  [None] has len of 1 and then returning matches[0] then returns a None.  And the consumers handle the None.  I do agree it's not an ideal way to implement the logic and I can come up with something simpler.

I like the ideas of propagating an exception.  The the upstream code is not prepared for that right now and we'd want all branches in this function to do that.

For example, the code that gets the thumbnails does get None back when searching for track images (happens in the old implementation as well - I'm not sure why that happens yet, but its on my list) and it happily continues along (minus the track image showing in the media browser). An exception propagating would cause the browse to fail and the track list wouldn't be shown.

https://github.com/home-assistant/core/blob/9580adfde90b04235bd338a45cf6c147286a2ae4/homeassistant/components/sonos/media_browser.py#L57-L62

We could change the function contract to use exceptions; and I'd like to do that in play_media also so that when media fails to play it gets propagated back (right now it logs errors).  But that may best be done in a separate pull request?

Reply:
Totally fair. How about just changing the return typing to `MusicServiceItem | None` if all consumers can already handle this?

Reply:
Looks like my statement was not accurate.  There is one spot that does not check the result of get_media, so when I change it I get a linter error.

Error: homeassistant/components/sonos/media_player.py:632: error: Item "None" of "Any | None" has no attribute "get_uri"  [union-attr]

https://github.com/home-assistant/core/blob/345fa755625713f6f8e09bb47a9a21b1f388d47f/homeassistant/components/sonos/media_player.py#L627-L631

Now, I have fixed that section of code in #113934 - to check for None and log an error.

Reply:
I reverted that change for now, looks like there's a breakage in dev that's preventing ruff from being happy on tests/components/light/common.py

Reply:
@jjlawren - we got some choices here in how to resolve this comment:

a) Get pull request #113934 reviewed and merged first, then we can rebase this and make the change to return MusicServiceItem | None - which will then pass lint.  This PR already needs #109518 to be merged and then rebased.

b) Don't make then MusicServiceItem | None change in this pull request.  I'll create a new pull request with just that change and once this PR and #113934 are merged; we can push that through

c) Combine the changes from #113934 into this pull request.

Option (c) - makes this pull request large and "fix more than one thing" - thoughts?

I have no preference on A or B; it just sequencing these changes.  Thoughts?

Reply:
Let's go with (a), that feels cleanest. #109518 has been merged, will look at reviewing #113934 soon.

Reply:
I have (a) in and merged.  Back to the state where the unrelated test for the switch fails only in CI.